src/chain/shared_state.rs

```diff
@@ -156,6 +157,32 @@ impl SharedState {
                 .collect()
         }
     }
+
+    /// Updates the entry in `their_keys` for `prefix` to the latest known key; if a split
+    /// occurred in the meantime, the keys for sections covering the rest of the address space are
+    /// initialised to the old key that was stored for their common ancestor
+    pub fn update_their_keys(&mut self, prefix: Prefix<XorName>, key: BlsPublicKey) {
+        if let Some(&pfx) = self
+            .their_keys
+            .keys()
+            .find(|pfx| pfx.is_compatible(&prefix))
+        {
+            let old_key = unwrap!(self.their_keys.remove(&pfx));
+            let old_pfx_sibling = pfx.sibling();
+            let mut current_pfx = prefix.sibling();
+            while !self.their_keys.contains_key(&current_pfx) && current_pfx != old_pfx_sibling {
+                let _ = self.their_keys.insert(current_pfx, old_key.clone());
+                current_pfx = current_pfx.popped().sibling();
```

Comment:
why a sibling after pop ?
is a `pop` enough?

Reply:
Because we want to cover the whole address space that was covered by the old prefix.

Say we have a key for prefix `1` and receive one for prefix `1000`. This means that `1` split at some point, but we don't know what the other sections are. We decide to just cover the space of prefix `1` with shortest possible prefixes, such that one of them is `1000` - this means we have to create entries for prefixes `1000`, `1001`, `101` and `11`. So we have:

* `1000` → sibling → `1001`
* `1001` → popped → `100` → sibling → `101`
* `101` → popped → `10` → sibling → `11`

Reply:
hmm... 
does the `with shortest possible prefixes` from `1000` to `1` is : `1000` -> `100` -> `10` -> `1` ?

Reply:
But we don't want to just go from `1000` to `1`. We want prefixes that cover the whole address space that `1` was covering. Basically, we want a set of potential prefixes that `1` could have split into, provided that one of them is `1000`.

If prefix `1000` exists on the network, then neither `1`, nor `10`, nor `100` can exist. But `11` can exist, and so can `101` and `1001`.